source/client-side-encryption/client-side-encryption.rst

```diff
@@ -2302,6 +2311,15 @@ key vault collection, the key management functions may be overloaded to take an
 explicit session parameter as described in the
 `Drivers Sessions Specification <../sessions/driver-sessions.rst>`__.
 
+Why is IndexKeyId needed?
+-------------------------
+
+Queryable Encryption supports two keys for an encrypted value: UserKey and IndexKey.
+- The UserKey encrypts user data. 
+- The IndexKey is used to derive tokens for indexing. It must match the "keyId" in encryptedFields for correct query results.
+
+Adding an EncryptOpts.indexKeyId enables multiple values with different UserKey to be inserted on the same field.
```

Comment:
@markbenvenuto is this rationale correct? If IndexKeyId is not necessary, I would like to leave it out of the API.

Reply:
You could remove it IF you do the analysis on the client-side to find the indexKeyId from the encryptedFields in collection metadata. The server will have to do the analysis to check they are the same.

You could let the user leave it blank and then fill it in for the user based on encryptedFields.

Reply:
> You could remove it IF you do the analysis on the client-side to find the indexKeyId from the encryptedFields in collection metadata. The server will have to do the analysis to check they are the same.

That may be non trivial. `ClientEncryption::Encrypt` does not know which field the payload will be applied to. To determine the field, the client would need to do query analysis.

Example with Go:

```go
// encryptedFields on server uses keyID1 on field "foo".
eo := options.Encrypt().
    SetAlgorithm("Indexed").
    SetKeyID(key2ID).
    SetIndexKeyID(key1ID)
// Encrypt() does not know which field insertPayload will be applied to.
insertPayload, err := clientEncryption.Encrypt(context.Background(), rawVal, eo)
assert.Nil(mt, err, "error on encrypt: %v", err)

coll := encryptedClient.Database("db").Collection("explicit_encryption")
_, err = coll.UpdateOne(context.Background(), bson.M{}, bson.M{"$set": bson.M{"encryptedIndexed": insertPayload}})
assert.Nil(mt, err, "error in UpdateOne: %v", err)
```

The default behavior of libmongocrypt is to assume IndexKeyId == UserKeyId. AFAICT adding an IndexKeyId option only enables the use-case for inserting data with IndexKeyId != UserKeyId. Is inserting with IndexKeyId != UserKeyId is a use-case we need to support?